

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tombo Commands &mdash; Tombo 1.4 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Re-squiggle Algorithm" href="resquiggle.html" />
    <link rel="prev" title="Welcome to Tombo’s documentation!" href="index.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tombo Commands</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#re-squiggle-raw-signal-genomic-alignment">Re-squiggle (Raw Signal Genomic Alignment)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#modified-base-detection">Modified Base Detection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#specific-alternative-base-method">Specific Alternative Base Method</a></li>
<li class="toctree-l3"><a class="reference internal" href="#de-novo-non-canonical-base-method"><em>De novo</em> Non-canonical Base Method</a></li>
<li class="toctree-l3"><a class="reference internal" href="#canonical-sample-comparison-method">Canonical Sample Comparison Method</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#text-output">Text Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#genome-browser-file-output">Genome Browser File Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genome-sequence-output">Genome Sequence Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#plotting-commands">Plotting Commands</a></li>
<li class="toctree-l2"><a class="reference internal" href="#read-filtering">Read Filtering</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="text_output.html">Text Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="rna.html">RNA Processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_training.html">Model Training (Advanced Users Only)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Tombo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Tombo Commands</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tombo-commands">
<h1>Tombo Commands<a class="headerlink" href="#tombo-commands" title="Permalink to this headline">¶</a></h1>
<p>Below are minimal use case examples. For more detail on each commands’ options and further algorithm details, please see the corresponding documentation sections.</p>
<div class="section" id="re-squiggle-raw-signal-genomic-alignment">
<h2>Re-squiggle (Raw Signal Genomic Alignment)<a class="headerlink" href="#re-squiggle-raw-signal-genomic-alignment" title="Permalink to this headline">¶</a></h2>
<p>The re-squiggle algorithm aligns raw signal (electric current nanopore measurements) to genomic/transcriptomic sequence.</p>
<p>One of the major assumptions of the re-squiggle algorithm is that the provided reference sequence is correct. Thus for poorly assembled genomes or divergent samples, an assembly polishing step (possibly from the same data/sample) may improve results.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">resquiggle</span></code> command will add infomation (the mapped genomic location and the raw signal to sequence assignment) to the read files provided (in FAST5 format), as well as producing an index file for more efficient file access in downstream commands.</p>
<div class="admonition important">
<p class="first admonition-title">Important</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">resquiggle</span></code> command must be run before any further processing by Tombo commands.</p>
</div>
<p><strong>Note</strong>: Tombo currenly includes default canonical models for both DNA or RNA data (supporting R9.4 and R9.5; 1D and 1D^2; R9.*.1 chemistries). Analysis of other nanopore data types is not supported at this time (e.g. R7 data). If DNA or RNA sample type is not explicitly specified (via <code class="docutils literal notranslate"><span class="pre">--dna</span></code> or <code class="docutils literal notranslate"><span class="pre">--rna</span></code> options) the sample type will be detected automatically from the raw read files.</p>
<p>For more details see the <a class="reference internal" href="resquiggle.html"><span class="doc">re-squiggle documentation</span></a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># annotate raw FAST5s with FASTQ files produced from the same reads</span>
<span class="c1"># skip this step if raw read files already contain basecalls</span>
tombo annotate_raw_with_fastqs --fast5-basedir &lt;fast5s-base-directory&gt; --fastq-filenames &lt;reads.fastq&gt;

tombo resquiggle &lt;fast5s-base-directory&gt; &lt;reference-fasta&gt; --processes <span class="m">4</span> --num-most-common-errors <span class="m">5</span>
</pre></div>
</div>
</div>
<div class="section" id="modified-base-detection">
<h2>Modified Base Detection<a class="headerlink" href="#modified-base-detection" title="Permalink to this headline">¶</a></h2>
<p>Tombo provides three methods for the investigation of modified bases (within the <code class="docutils literal notranslate"><span class="pre">detect_modifications</span></code> command group). Each method has different advantages and requirements.</p>
<hr class="docutils" />
<div class="figure align-center" id="id1">
<img alt="_images/testing_method_comparison.png" src="_images/testing_method_comparison.png" />
<p class="caption"><span class="caption-text">Tombo modified base testing methods.</span></p>
</div>
<hr class="docutils" />
<p>All modified base detection methods poduce per-read, per-genomic position test statistics (which can be saved via the <code class="docutils literal notranslate"><span class="pre">--per-read-statistics-basename</span></code> option). A threshold is then applied to these statistics to produce an estimate for the fraction of reads that appear modified at each genomic location.</p>
<ol class="arabic simple">
<li><strong>Specific alternative base detection (recommended)</strong></li>
</ol>
<blockquote>
<div><ul class="simple">
<li>Run using <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">detect_modifications</span> <span class="pre">alternative_model</span></code> command.</li>
<li>This method identifies signal that deviates from the canonical base expected signal level while matching a specific alternative base expected signal level.</li>
<li>This method produces a statistic similar to a log likelihood ratio, but scaled to be more robust to outlier signal assignments (similar to <a class="reference external" href="http://mathworld.wolfram.com/TukeysBiweight.html">Tukey’s biweight function</a>).</li>
<li>Alternative DNA models are currently available for 5-methylcytosine (5mC) and N6-methyladenosine (6mA) in all sequence contexts.</li>
<li>An alternative RNA model is available for 5mC.</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><strong>De novo canonical model comparison</strong></li>
</ol>
<blockquote>
<div><ul class="simple">
<li>Run using <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">detect_modifications</span> <span class="pre">de_novo</span></code> command.</li>
<li>This method compares re-squiggled signal to the default canonical model.</li>
<li>While this method may produce significant false positive and negative results per-read, it produces the best results for many statistical measures per-genomic location (fraction of modified bases across a set of reads).</li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li><strong>Canonical (control) sample comparison</strong></li>
</ol>
<blockquote>
<div><ul class="simple">
<li>Run using <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">detect_modifications</span> <span class="pre">sample_compare</span></code> command.</li>
<li>This method performs a hypothesis test against the distribution estimated from the control sample at each base.</li>
<li>This method requires the production of a second set of reads containing only the 4 canonical bases (e.g PCR for DNA of IVT for RNA).</li>
</ul>
</div></blockquote>
<blockquote>
<div>Both the control sample comparison and the <em>de novo</em> methods may not identify the exact modified base location (as the shifted signal does not always center on a modified base) and gives no information as to the identity of a modified base.</div></blockquote>
<p>The result of all <code class="docutils literal notranslate"><span class="pre">detect_modifications</span></code> calls will be a binary statistics file(s), which can be passed to other Tombo commands.</p>
<p>For more details see the <a class="reference internal" href="modified_base_detection.html"><span class="doc">modified base detection documentation</span></a>.</p>
<div class="section" id="specific-alternative-base-method">
<h3>Specific Alternative Base Method<a class="headerlink" href="#specific-alternative-base-method" title="Permalink to this headline">¶</a></h3>
<p>In order to specifically detect 5mC and 6mA, use the <code class="docutils literal notranslate"><span class="pre">detect_modifications</span> <span class="pre">alternative_model</span></code> command.</p>
<p>This will compute a statistic similar to a log likelihood ratio using the default canonical model and the 5mC and 6mA alternative DNA models provided with Tombo.</p>
<p>This is the perferred method for modified base detection if a model is available for your biological modification of interest, as it identifies the exact location of the modified base and reduces false positives for spurious shifts in signal.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo detect_modifications alternative_model --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --alternate-bases 5mC 6mA --statistics-file-basename sample_alt_model
</pre></div>
</div>
</div>
<div class="section" id="de-novo-non-canonical-base-method">
<h3><em>De novo</em> Non-canonical Base Method<a class="headerlink" href="#de-novo-non-canonical-base-method" title="Permalink to this headline">¶</a></h3>
<p>In order to perform <em>de novo</em> non-canonical base detection, use the <code class="docutils literal notranslate"><span class="pre">detect_modifications</span> <span class="pre">de_novo</span></code> command.</p>
<p>This will perform a hypothesis test against the default canonical base model provided with Tombo. Note that this method is quite error prone and may result in a high false positive rate on a per-read basis, but may be of use in a research and development setting. This method also has the lowest requirement, consisting of only a set of reads potentially containing modifications and a reference sequence.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo detect_modifications de_novo --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --statistics-file-basename sample_de_novo_detection
</pre></div>
</div>
</div>
<div class="section" id="canonical-sample-comparison-method">
<h3>Canonical Sample Comparison Method<a class="headerlink" href="#canonical-sample-comparison-method" title="Permalink to this headline">¶</a></h3>
<p>In order to execute the canonical sample comparison method, use the <code class="docutils literal notranslate"><span class="pre">detect_modifications</span> <span class="pre">sample_compare</span></code> command.</p>
<p>This will perform a hypothesis test against the signal level observed from the control sample (provided via <code class="docutils literal notranslate"><span class="pre">--control-fast5-basedirs</span></code> option) at each genomic position. This method (like the <code class="docutils literal notranslate"><span class="pre">de_novo</span></code> method) does not always identify the exact modified base position.</p>
<p>As of version 1.4, this method uses the canonical base model as a prior for control sample distribution estimation drastically improving results, particularly for low coverage samples. To test only against the canonical sample use the <code class="docutils literal notranslate"><span class="pre">--sample-only-estimates</span></code> flag. The prior weights for the estimated mean and standard deviation can be set using the <code class="docutils literal notranslate"><span class="pre">--model-prior-weights</span></code> option.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo detect_modifications sample_compare --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --control-fast5-basedirs &lt;control-fast5s-base-directory&gt; <span class="se">\</span>
    --statistics-file-basename sample_canonical_compare
</pre></div>
</div>
</div>
</div>
<div class="section" id="text-output">
<h2>Text Output<a class="headerlink" href="#text-output" title="Permalink to this headline">¶</a></h2>
<div class="section" id="genome-browser-file-output">
<h3>Genome Browser File Output<a class="headerlink" href="#genome-browser-file-output" title="Permalink to this headline">¶</a></h3>
<p>In order to output the results of re-squiggling and statistical testing in a genome browser compatible format (either <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/wiggle.html">wiggle format</a> or <a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/bedgraph.html">bedgraph format</a>), the <code class="docutils literal notranslate"><span class="pre">tombo</span> <span class="pre">text_output</span> <span class="pre">genome_browser</span></code> command is provided.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo text_output browser_files --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --statistics-filename sample_alt_model.5mC.tombo.stats <span class="se">\</span>
     --browser-file-basename sample_alt_model --file-types dampened_fraction coverage
</pre></div>
</div>
<div class="admonition hint">
<p class="first admonition-title">Hint</p>
<p>Other <code class="docutils literal notranslate"><span class="pre">--file-types</span></code> available are <code class="docutils literal notranslate"><span class="pre">fraction</span></code>, <code class="docutils literal notranslate"><span class="pre">valid_coverage</span></code>, <code class="docutils literal notranslate"><span class="pre">signal</span></code>, <code class="docutils literal notranslate"><span class="pre">signal_sd</span></code>, <code class="docutils literal notranslate"><span class="pre">dwell</span></code> and <code class="docutils literal notranslate"><span class="pre">difference</span></code>.</p>
<p class="last">The <code class="docutils literal notranslate"><span class="pre">dampened_fraction</span></code> option adds psuedo-counts to the detected number of un-modified and modified reads at each tested location (as specified by the <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> option), while the <code class="docutils literal notranslate"><span class="pre">fraction</span></code> option returns the raw fraction of modified reads at any genomic site from <code class="docutils literal notranslate"><span class="pre">detect_modifications</span></code> results. The <code class="docutils literal notranslate"><span class="pre">dampen_fraction</span></code> output is intended to allow the inclusion of low coverage regions in downstream analysis without causing potentially false site at the top of rank lists. Visualize different values of the <code class="docutils literal notranslate"><span class="pre">--coverage-dampen-counts</span></code> option with the included <code class="docutils literal notranslate"><span class="pre">scripts/test_beta_priors.R</span></code> script.</p>
</div>
</div>
<div class="section" id="genome-sequence-output">
<h3>Genome Sequence Output<a class="headerlink" href="#genome-sequence-output" title="Permalink to this headline">¶</a></h3>
<p>For modified base analysis pipelines (e.g. motif detection), it may be useful to output the genomic sequence surrounding locations with the largest fraction of modified reads. The <code class="docutils literal notranslate"><span class="pre">text_output</span> <span class="pre">signif_sequence_context</span></code> command is provided for this purpose.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo text_output signif_sequence_context --statistics-filename sample_alt_model.6mA.tombo.stats <span class="se">\</span>
    --genome-fasta &lt;reference-fasta&gt; --sequences-filename sample_alt_model.6mA.most_signif.fasta
</pre></div>
</div>
<p>Example <a class="reference external" href="http://meme-suite.org/doc/meme.html">meme</a> command line modified base motif detection command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./meme -oc motif_output.meme -dna -mod zoops sample_alt_model.6mA.most_signif.fasta
</pre></div>
</div>
<p>For more details see the <a class="reference internal" href="text_output.html"><span class="doc">text output documentation</span></a>.</p>
</div>
</div>
<div class="section" id="plotting-commands">
<h2>Plotting Commands<a class="headerlink" href="#plotting-commands" title="Permalink to this headline">¶</a></h2>
<p>Tombo provides many plotting functions for the visualization of modified bases and raw nanopore signal in general.</p>
<p>Most plotting commands are genome-anchored. That is the raw signal is plotted as the re-squiggle algorithm has assigned it to the genome. Thus each read may contain a different number of raw observations assigned to each genomic base. For regions with higher coverage, several over-plotting options are available. For those options producing a distribution, these are taken over each reads average signal assigned to a base. This requires extraction of these levels from all relevant FAST5 files and thus can be slow for very deep coverage regions.</p>
<p>Each genome anchored plotting command allows for the selection of genomic positions based on generally applicable criterion.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tombo plot max_coverage --fast5-basedirs &lt;fast5s-base-directory&gt; --plot-standard-model

tombo plot motif_centered --fast5-basedirs &lt;fast5s-base-directory&gt; --motif AWC <span class="se">\</span>
    --genome-fasta genome.fasta --control-fast5-basedirs &lt;control-fast5s-base-directory&gt;

tombo plot per_read --per-read-statistics-filename &lt;per_read_statistics.tombo.stats&gt; <span class="se">\</span>
    --genome-locations chromosome:1000 chromosome:2000:- <span class="se">\</span>
    --genome-fasta genome.fasta
</pre></div>
</div>
<p>For more details see the <a class="reference internal" href="plotting.html"><span class="doc">plotting documentation</span></a>.</p>
</div>
<div class="section" id="read-filtering">
<h2>Read Filtering<a class="headerlink" href="#read-filtering" title="Permalink to this headline">¶</a></h2>
<p>Read filtering commands can be useful to extract the most out out of a set of reads for modified base detection. Read filtering commands effect only the Tombo index file, and so filters can be cleared or applied iteratively without re-running the re-squiggle command. Five filters are currently made available (<code class="docutils literal notranslate"><span class="pre">genome_locations</span></code>, <code class="docutils literal notranslate"><span class="pre">raw_signal_matching</span></code>, <code class="docutils literal notranslate"><span class="pre">q_score</span></code>, <code class="docutils literal notranslate"><span class="pre">level_coverage</span></code> and <code class="docutils literal notranslate"><span class="pre">stuck</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># filter reads to a specific genomic location</span>
tombo filter genome_locations --fast5-basedirs path/to/native/rna/fast5s/ <span class="se">\</span>
    --include-regions chr1:0-10000000

<span class="c1"># apply a more strigent observed to expected signal score (default: 1.1 for DNA reads)</span>
tombo filter raw_signal_matching --fast5-basedirs path/to/native/rna/fast5s/ <span class="se">\</span>
    --signal-matching-score <span class="m">1</span>.0
</pre></div>
</div>
<div class="admonition hint">
<p class="first admonition-title">Hint</p>
<p class="last">Hint: Save a set of filters for later use by copying the Tombo index file: <code class="docutils literal notranslate"><span class="pre">cp</span> <span class="pre">path/to/native/rna/.fast5s.RawGenomeCorrected_000.tombo.index</span> <span class="pre">save.native.tombo.index</span></code>. To re-set to a set of saved filters after applying further filters simply replace the index file: <code class="docutils literal notranslate"><span class="pre">cp</span> <span class="pre">save.native.tombo.index</span> <span class="pre">path/to/native/rna/.fast5s.RawGenomeCorrected_000.tombo.index</span></code>.</p>
</div>
<blockquote>
<div>For more read filtering commands see the <a class="reference external" href="https://nanoporetech.github.io/tombo/filtering.html">Tombo filter documentation here</a>.</div></blockquote>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">For additional command details, see the specific commands documentation section.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="resquiggle.html" class="btn btn-neutral float-right" title="Re-squiggle Algorithm" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Welcome to Tombo’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-18, Oxford Nanopore Technologies.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.4',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>